Mathlib/Data/Matroid/Dual.lean

```diff
@@ -0,0 +1,264 @@
+      refine' (subset_union_of_subset_right (subset_diff.mpr ⟨hIB',_⟩) _).trans hI'B₁
+      refine' disjoint_of_subset_left hI.subset disjoint_sdiff_left
+
+    subst this
+
+    refine' subset_diff.mpr ⟨hJX, by_contra (fun hne ↦ _)⟩
+    obtain ⟨e, heJ, heB'⟩ := not_disjoint_iff.mp hne
+    obtain (heB'' | ⟨-,heX⟩ ) := hB₁I heB'
+    · exact hdj.ne_of_mem heJ heB'' rfl
+    exact heX (hJX heJ)
+  subset_ground := by tauto
+
+/-- The dual of a matroid; the bases are the complements (w.r.t `M.E`) of the bases of `M`. -/
+def dual (M : Matroid α) : Matroid α := M.dualIndepMatroid.matroid
+
+/-- The `﹡` symbol, which denotes matroid duality.
+  (This is distinct from the usual `*` symbol for multiplication, due to precedence issues. )-/
+postfix:max "﹡" => Matroid.dual
+
+theorem dual_indep_iff_exists' : (M﹡.Indep I) ↔ I ⊆ M.E ∧ (∃ B, M.Base B ∧ Disjoint I B) := by
+  simp [dual]
+
+@[simp] theorem dual_ground : M﹡.E = M.E := rfl
+
+@[simp] theorem dual_indep_iff_exists (hI : I ⊆ M.E := by aesop_mat) :
+    M﹡.Indep I ↔ (∃ B, M.Base B ∧ Disjoint I B) := by
+  rw [dual_indep_iff_exists', and_iff_right hI]
+
+theorem dual_dep_iff_forall : (M﹡.Dep I) ↔ (∀ B, M.Base B → (I ∩ B).Nonempty) ∧ I ⊆ M.E := by
+  simp_rw [dep_iff, dual_indep_iff_exists', dual_ground, and_congr_left_iff, not_and,
+    not_exists, not_and, not_disjoint_iff_nonempty_inter, imp_iff_right_iff, iff_true_intro Or.inl]
+
+instance dual_finite [M.Finite] : M﹡.Finite :=
+  ⟨M.ground_finite⟩
+
+instance dual_nonempty [M.Nonempty] : M﹡.Nonempty :=
+  ⟨M.ground_nonempty⟩
+
+@[simp] theorem dual_base_iff (hB : B ⊆ M.E := by aesop_mat) : M﹡.Base B ↔ M.Base (M.E \ B) := by
+  rw [base_compl_iff_mem_maximals_disjoint_base, base_iff_maximal_indep, dual_indep_iff_exists',
+    mem_maximals_setOf_iff]
+  simp [dual_indep_iff_exists']
+
+theorem dual_base_iff' : M﹡.Base B ↔ M.Base (M.E \ B) ∧ B ⊆ M.E :=
+  (em (B ⊆ M.E)).elim (fun h ↦ by rw [dual_base_iff, and_iff_left h])
+    (fun h ↦ iff_of_false (h ∘ (fun h' ↦ h'.subset_ground)) (h ∘ And.right))
+
+theorem setOf_dual_base_eq : {B | M﹡.Base B} = (fun X ↦ M.E \ X) '' {B | M.Base B} := by
+  ext B
+  simp only [mem_setOf_eq, mem_image, dual_base_iff']
+  refine' ⟨fun h ↦ ⟨_, h.1, diff_diff_cancel_left h.2⟩,
+    fun ⟨B', hB', h⟩ ↦ ⟨_,h.symm.trans_subset (diff_subset _ _)⟩⟩
+  rwa [←h, diff_diff_cancel_left hB'.subset_ground]
+
+@[simp] theorem dual_dual (M : Matroid α) : M﹡﹡ = M :=
+  eq_of_base_iff_base_forall rfl (fun B (h : B ⊆ M.E) ↦
+    by rw [dual_base_iff, dual_base_iff, dual_ground, diff_diff_cancel_left h])
+
+theorem dual_involutive : Function.Involutive (dual : Matroid α → Matroid α) := dual_dual
+
+theorem dual_injective : Function.Injective (dual : Matroid α → Matroid α) :=
+  dual_involutive.injective
+
+@[simp] theorem dual_inj_iff {M₁ M₂ : Matroid α} : M₁﹡ = M₂﹡ ↔ M₁ = M₂ :=
+  dual_injective.eq_iff
+
+theorem dual_inj {M₁ M₂ : Matroid α} (h : M₁﹡ = M₂﹡) : M₁ = M₂ :=
+  dual_inj_iff.1
```

Comment:
This lemma is defeq to `dual_injective`; I don't think there's any need to have both.